### PR TITLE
Update ECAL and HCAL reconstruction to run on multple GPUs

### PR DESCRIPTION
**Pull request by vkhristenko**
#### PR description:

superseeds #498 

hcal and ecal are done together in here cause of the change `CUDADataFormats/HcalCommon` -> `CUDADataFormats/CaloCommon` which both hcal and ecal now depend on. this avoids duplication...

this is to allow hcal and ecal running on a node with multiple gpus.
all the modules have been updated for that and now basically no protection for cuda service is needed.
__note: Ecal RecHit was only updated but not validated (fillDescriptions is not self-sufficient)__ @amassiro  

the only thing in this pr is that the newly added conditions' Records should be moved to CondFormats/DataRecord eventually.

#### PR validation:

using standalone execs

### CODE REVIEW

File: RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu [hunk 1]
```diff
@@ -247,8 +263,8 @@ namespace ecal {
           // likely false
           if (check_hasSwitchToGain0) {
             // assign for the case some sample having gainId == 0
-            //energies[ch] = amplitudes[ch][sample_max];
-            energies[ch] = amplitude;
+            //energies[inputCh] = amplitudes[ch][sample_max];
+            energies[inputCh] = amplitude;
```
Comment (fwyzard) [inline RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu:267]: why the change from `ch` to `inputCh` ?
what is the difference ?

File: RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu [hunk 2]
```diff
@@ -65,6 +71,16 @@ namespace ecal {
...
+      // need to ref the right ptr
+      // macro is for clarity and safety
+#define ARRANGE(var) auto *var = ch >= offsetForInputs ? var##EE : var##EB
+      ARRANGE(amplitudesForMinimization);
+      ARRANGE(energies);
+      ARRANGE(chi2);
+      ARRANGE(g_pedestal);
+      ARRANGE(dids_out);
+      ARRANGE(flags);
+
```
Comment (fwyzard) [inline RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu:83]: could you `#undef ARRANGE` once you are done using it ?

File: RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu [hunk 3]
```diff
@@ -348,12 +358,12 @@ namespace ecal {
 
         // store to global output values
         // FIXME: amplitudes are used in global directly
-        chi2s[idx] = chi2;
-        energies[idx] = resultAmplitudes(5);
+        chi2s[inputCh] = chi2;
+        energies[inputCh] = resultAmplitudes(5);
```
Comment (fwyzard) [inline RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu:362]: same question here: what is the difference between `idx` and `inputCh` ?

File: RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu (hunk 1 above)
Comment (vkhristenko) [reply RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu:267]: previously there was a single ptr for the output collections (eb was first and then ee). now i switched to 2 ptrs, one for each collection given that each collection has its own unique_ptr that manages memory. `inputCh` is computed having this in mind... ch is just a linear idx.

File: RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu (hunk 3 above)
Comment (vkhristenko) [reply RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationKernels.cu:363]: same as above

File: RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu (hunk 1 above)
Comment (fwyzard) [reply RecoLocalCalo/EcalRecProducers/plugins/AmplitudeComputationCommonKernels.cu:268]: understood, thanks

File: EventFilter/EcalRawToDigi/plugins/EcalRawToDigiGPU.cc [hunk 4]
```diff
@@ -97,6 +85,40 @@ void EcalRawToDigiGPU::acquire(edm::Event const& event,
...
   event.getByToken(rawDataToken_, rawDataHandle);
 
+  // scratch
+  ecal::raw::ScratchDataGPU scratchGPU = {
+    cms::cuda::make_device_unique<uint32_t[]>(2, ctx.stream())
+  };
+
+  // input cpu data
+  ecal::raw::InputDataCPU inputCPU = {
+    cms::cuda::make_host_unique<unsigned char[]>(
```
Comment (fwyzard) [inline EventFilter/EcalRawToDigi/plugins/EcalRawToDigiGPU.cc:95]: comment for the future (after CMSSW 11.1.0 / CUDA 11.0 / c++17): would it make sense to use `std::byte` instead of `unsigned char` ?

Comment (vkhristenko) [reply EventFilter/EcalRawToDigi/plugins/EcalRawToDigiGPU.cc:95]: I would say we should be consistent with what sits inside of the FEDRawData... although we do make a copy before sending to the device...

File: RecoLocalCalo/HcalRecProducers/src/DeclsForKernels.h [hunk 5]
```diff
@@ -75,26 +82,28 @@ namespace hcal {
...
+          config.maxChannels, cudaStream);
       }
     };
 
     struct ScratchDataGPU {
+      cms::cuda::device::unique_ptr<float[]> amplitudes, noiseTerms,
+          pulseMatrices, pulseMatricesM, pulseMatricesP;
+      cms::cuda::device::unique_ptr<int8_t[]> soiSamples;
+
+      /*
```
Comment (fwyzard) [inline RecoLocalCalo/HcalRecProducers/src/DeclsForKernels.h:106]: Could you delete these, or there is some use in keeping them ?

Comment (vkhristenko) [reply RecoLocalCalo/HcalRecProducers/src/DeclsForKernels.h]: addressed

Comment (fwyzard) [reply RecoLocalCalo/HcalRecProducers/src/DeclsForKernels.h]: Thanks

**Comment by fwyzard** [comment]
Sorry for taking long to go through this, my plan is to
  - finish review the changes
  - check that it works fine with 2 GPUs

I am also using this PR as an excuse to add the HCAL-only workflows to the validation, but if that takes too long I'll just go ahead and merge it.

**Comment by vkhristenko** [comment]
@fwyzard np. i tested this on cmg-gpu1080...

**Comment by fwyzard** [comment]
I haven't gotten to the HCAL workflow yet - but this PR already breaks the ECAL-only workflow, and the HLT customisation.

Looks like the reasons are:
  - the old parameters left in the HLT workflow
  - the new ESProdcers missing from the HLT and offline workflows

I can try to add them ...

**Comment by fwyzard** [comment]
What is the configuration for `EcalRecHitParametersGPUESProducer` supposed to look like ?

The autogenerated cfi is
```python
import FWCore.ParameterSet.Config as cms

ecalRecHitParametersGPUESProducer = cms.ESSource('EcalRecHitParametersGPUESProducer',
  ChannelStatusToBeExcluded = cms.VPSet(
    cms.PSet()
  ),
  appendToDataLabel = cms.string('')
)
```
but using it fails with

> ----- Begin Fatal Exception 08-Jul-2020 19:06:29 CEST-----------------------
> An exception of category 'Configuration' occurred while
>    [0] Constructing the EventProcessor
>    [1] Validating configuration of ESProducer or ESSource of type EcalRecHitParametersGPUESProducer with label: 'ecalRecHitParametersGPUESProducer'
> Exception Message:
> Missing required parameter.  It should have label "kDAC" and have type "tracked string".
> The description has no default.  The parameter must be defined in the configuration
> ----- End Fatal Exception -------------------------------------------------

**Comment by vkhristenko** [comment]
The auto generated cfi is not self sufficient...I indicated this in pr
description. Andrea should provide input here.

**Comment by fwyzard** [comment]
I see - I hadn't understood that the comment was about the EDProducer/ESProducer split, thanks for the clarification.

**Comment by vkhristenko** [comment]
> I see - I hadn't understood that the comment was about the EDProducer/ESProducer split, thanks for the clarification.

ok, I think that you need to use this guy https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_1_X_Patatrack/RecoLocalCalo/EcalRecProducers/python/ecalRecHitGPU_cfi.py

but again the EDProducer itself does not really have self-sufficient `fillDescriptions`. And when i moved to ES..., I moved only what was present in the EDProducer. Therefore, I think somewhere in the customizations there should be loading of this cfi... @amassiro

**Comment by amassiro** [comment]
> > I see - I hadn't understood that the comment was about the EDProducer/ESProducer split, thanks for the clarification.
> 
> ok, I think that you need to use this guy https://github.com/cms-patatrack/cmssw/blob/CMSSW_11_1_X_Patatrack/RecoLocalCalo/EcalRecProducers/python/ecalRecHitGPU_cfi.py
> 
> but again the EDProducer itself does not really have self-sufficient `fillDescriptions`. And when i moved to ES..., I moved only what was present in the EDProducer. Therefore, I think somewhere in the customizations there should be loading of this cfi... @amassiro




Thanks @vkhristenko !

I think that with this 
https://github.com/cms-patatrack/cmssw/pull/504
the configuration producer should be ok.

I tested with the usual test configuration under RecoLocalCalo/EcalRecProducers/test/testEcalRechitProducer_cfg.py

**Comment by fwyzard** [comment]
Validated together with #504, #507 and #508.
